llvm/tools/sycl-post-link/ModuleSplitter.cpp

```diff
@@ -200,93 +171,142 @@ groupEntryPointsByKernelType(ModuleDesc &MD,
-    return make_range(AddrTakenFunctions.begin(), AddrTakenFunctions.end());
+private:
+  void addUserToGraphRecursively(const User *Root, const GlobalValue *V) {
+
+    SmallVector<const User *, 8> WorkList;
+    WorkList.push_back(Root);
+
+    while (!WorkList.empty()) {
+      const User *U = WorkList.pop_back_val();
+      if (const auto *I = dyn_cast<const Instruction>(U)) {
+        const auto *UFunc = I->getFunction();
+        Graph[UFunc].insert(V);
+      } else if (isa<const Constant>(U)) {
+        if (const auto *GV = dyn_cast<const GlobalVariable>(U))
+          Graph[GV].insert(V);
+        // This could be a global variable or some constant expression (like
+        // bitcast or gep). We trace users of this constant further to reach
+        // global objects they are used by and add them to the graph.
+        for (const auto *UU : U->users())
+          WorkList.push_back(UU);
+      } else {
+        llvm_unreachable("Unhandled type of function user");
+      }
+    }
   }
+
+  DenseMap<const GlobalValue *, GlobalSet> Graph;
+  SmallPtrSet<const GlobalValue *, 1> EmptySet;
 };
 
-void collectFunctionsToExtract(SetVector<const GlobalValue *> &GVs,
-                               const EntryPointGroup &ModuleEntryPoints,
-                               const CallGraph &Deps) {
+void collectFunctionsAndGlobalVariablesToExtract(
+    SetVector<const GlobalValue *> &GVs, const Module &M,
+    const EntryPointGroup &ModuleEntryPoints, const DependencyGraph &Deps,
+    const std::function<bool(const Function *)> &Filter = nullptr) {
+  // We start with module entry points
   for (const auto *F : ModuleEntryPoints.Functions)
     GVs.insert(F);
-  // It is conservatively assumed that any address-taken function can be invoked
-  // or otherwise used by any function in any module split from the initial one.
-  // So such functions along with the call graphs they start are always
-  // extracted (and duplicated in each split module). They are not treated as
-  // entry points, as SYCL runtime requires that intersection of entry point
-  // sets of different device binaries (for the same target) must be empty.
-  // TODO: try to determine which split modules really use address-taken
-  // functions and only duplicate the functions in such modules. Note that usage
-  // may include e.g. function address comparison w/o actual invocation.
-  for (const auto *F : Deps.addrTakenFunctions()) {
-    if (!isKernel(*F) && (isESIMDFunction(*F) == ModuleEntryPoints.isEsimd()))
-      GVs.insert(F);
+
+  // Non-discardable global variables are also include into the initial set
+  for (const auto &GV : M.globals()) {
+    if (!GV.isDiscardableIfUnused())
+      GVs.insert(&GV);
   }
 
   // GVs has SetVector type. This type inserts a value only if it is not yet
   // present there. So, recursion is not expected here.
   decltype(GVs.size()) Idx = 0;
   while (Idx < GVs.size()) {
```

Comment:
i might be missing something but can we not just use a range based loop here?

Reply:
We update the container as we iterate over it. This is a BFS which uses a single container as a queue, results storage and set of visited functions (to prevent recursion).